### PR TITLE
Add busy indicators (aka spinners)

### PR DESCRIPTION
**Pull request by nstrayer**
This PR adds busy indication, providing users with a visual cue when the server is busy calculating outputs or otherwise serving requests to the client. More specifically, a spinner is shown on each calculating/recalculating output, and a pulsing banner is shown at the top of the page when the app is otherwise busy. Use the new `ui.busy_indicator.options()` function to customize the appearance of the busy indicators and `ui.busy_indicator.use()` to disable/enable them.

Here is an illustration of the default spinner and pulse experience:

```shell
shiny run shiny/api-examples/busy_indicators/app-express.py
```

https://github.com/rstudio/shiny/assets/1365941/365cf231-0ed8-433d-93a3-ed7ce2e8aad8

This PR also adds `busyIndicatorOptions()` to customize the appearance of both spinners and the pulse. Here's an example showing off 4 of the main `spinner_type`s available: `ring`, `bars`, `dots`, and `pulse`.

```shell
shiny run examples/busy_indicators/app.py
```

https://github.com/rstudio/shiny/assets/1365941/52f056d4-2efe-4862-be54-7d7e16eed7f3

### CODE REVIEW

File: shiny/ui/_loading_spinners.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    Args:
+        `type`: The type of spinner to use. Options include "disc", "tadpole", and "dots".
+          Defaults to "tadpole".
+        `color`: The color of the spinner. This can be any valid CSS color. Defaults to
+          the current app "primary" color (if using a theme) or light-blue if not.
+        `size`: The size of the spinner. This can be any valid CSS size. Defaults to
+          "80px".
+        `speed`: The amount of time for the spinner to complete a single revolution. This
+          can be any valid CSS time. Defaults to "1s".
+    """
```
Comment (wch) [inline shiny/ui/_loading_spinners.py:31]: This should be formatted numpydoc-style. I believe that's necessary for quartodoc to work properly.

For example, see https://github.com/posit-dev/py-shiny/blob/1611d5f6e509b36b62e57bb7eaf1702bcd2effbe/shiny/ui/_input_text.py#L24-L64

File: shiny/ui/_loading_spinners.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return HTMLDependency(
+        "shiny-loading-spinners-css",
+        version=__version__,
+        source={
+            "package": "shiny",
+            "subdir": "www/shared/loading-spinners/",
+        },
+        stylesheet={"href": "spinners.css"},
+        all_files=True,
+        head="<style>" + rule_contents + "</style>",
```
Comment (wch) [inline shiny/ui/_loading_spinners.py:67]: What happens if someone includes two spinners of different types on a page?

**Comment by wch** [comment]
If the user experience here is smooth and not jarring, then I think it would make sense to enable spinners by default.

File: shiny/ui/_loading_spinners.py (hunk 2 above)
Comment (nstrayer) [reply shiny/ui/_loading_spinners.py:67]: Only the first one will be respected. I think that context would motivate the inclusion of the `with_spinner()` function in addition to the global spinners enabling

Comment (nstrayer) [reply shiny/ui/_loading_spinners.py:83]: At least first in terms of which html dependency was added to the page. I guess that _could_ occur out of logical order with dynamic ui etc..

**Review by nstrayer — COMMENTED** [review]
Updated the scope to enable spinners by default at the page-level. A small spinner is placed in the upper right corner of an app when any element on the page has `.recalculating`. There is also a way to enable spinners for a specific element or to turn them off by adding the class `no-spinner` to (anywhere on) the page.

File: shiny/ui/_html_deps_py_shiny.py [hunk 3]
```diff
@@ -52,3 +53,24 @@ def spin_dependency() -> HTMLDependency:
...
+
+def busy_indicators_dependency() -> HTMLDependency:
+    return HTMLDependency(
+        "shiny-busy-indicators",
+        __version__,
+        source={"package": "shiny", "subdir": "www/shared/py-shiny/busy-indicators"},
+        stylesheet={"href": "busy-indicators.css"},
+        head=TagList(
+            # Default to spinners mode
+            busy_indicators.mode("spinners"),
```
Comment (cpsievert) [inline shiny/ui/_html_deps_py_shiny.py:66]: If we merge this into this release, we'll need:

```suggestion
            # TODO: uncomment to default to spinners mode
            # busy_indicators.mode("spinners"),
```

File: shiny/ui/_html_deps_py_shiny.py [hunk 4]
```diff
@@ -52,3 +53,24 @@ def spin_dependency() -> HTMLDependency:
...
+        head=TagList(
+            # Enable busy indicators by default.
+            busy_indicators.use(),
+            # Show a page-level spinner up until the next idle.
+            # Note: this is only sensible when this dependency comes bundled with the
+            # main shiny dependency, which is currently the case (i.e., it should never
+            # come in through dynamic UI).
+            tags.script(
+                "document.documentElement.classList.add('shiny-not-yet-idle');"
+                + "$(document).on('shiny:idle', function() { document.documentElement.classList.remove('shiny-not-yet-idle'); });"
```
Comment (cpsievert) [inline shiny/ui/_html_deps_py_shiny.py:73]:
```suggestion
                + "$(document).one('shiny:idle', function() { document.documentElement.classList.remove('shiny-not-yet-idle'); });"
```

File: shiny/ui/_html_deps_py_shiny.py [hunk 5]
```diff
@@ -52,3 +53,24 @@ def spin_dependency() -> HTMLDependency:
...
+
+def busy_indicators_dependency() -> HTMLDependency:
+    return HTMLDependency(
+        "shiny-busy-indicators",
+        __version__,
+        source={"package": "shiny", "subdir": "www/shared/py-shiny/busy-indicators"},
+        stylesheet={"href": "busy-indicators.css"},
+        head=TagList(
+            # Enable busy indicators by default.
+            busy_indicators.use(),
```
Comment (cpsievert) [inline shiny/ui/_html_deps_py_shiny.py:66]: Could explore having something like a `html_attrs()` built on MetaDataNode that gets baked into htmltools

File: js/busy-indicators/busy-indicators.scss [hunk 6]
```diff
@@ -0,0 +1,181 @@
+/*
+  CSS for shiny.ui.busy_indicators module.
+
+  The general idea is to leverage the `recalculating` class that shiny.js adds to each
+  output when it's recalculating and add a spinner to the `::after` pseudo-element.
+*/
+
+/* "Built-in" spinner types */
+/* Part of the reason why they're all here is so we can embed the SVGs in the CSS */
```
Comment (wch) [inline js/busy-indicators/busy-indicators.scss:9]: Make it explicit in the comment that esbuild will embed them in the CSS (not just that the svg's can be embedded)

File: shiny/ui/_html_deps_py_shiny.py [hunk 7]
```diff
@@ -52,3 +53,24 @@ def spin_dependency() -> HTMLDependency:
...
+            # Show a page-level spinner up until the next idle.
+            # Note: this is only sensible when this dependency comes bundled with the
+            # main shiny dependency, which is currently the case (i.e., it should never
+            # come in through dynamic UI).
+            tags.script(
```
Comment (cpsievert) [inline shiny/ui/_html_deps_py_shiny.py:71]: Add a TODO that it would probably make to add this to shiny.js

File: shiny/ui/_html_deps_py_shiny.py [hunk 8]
```diff
@@ -52,3 +53,26 @@ def spin_dependency() -> HTMLDependency:
...
+        source={"package": "shiny", "subdir": "www/shared/py-shiny/busy-indicators"},
+        stylesheet={"href": "busy-indicators.css"},
+        # Add a CSS class to the root element up until the next idle. The
+        # busy-indicators.css above uses this class in order to indicate busy status
+        # when the app is loading.
+        # TODO: it'd be nice shiny.js did this
+        head=tags.script(
+            "document.documentElement.classList.add('shiny-not-yet-idle');"
+            + "$(document).one('shiny:idle', function() { document.documentElement.classList.remove('shiny-not-yet-idle'); });"
+        ),
```
Comment (cpsievert) [inline shiny/ui/_html_deps_py_shiny.py:71]: Put this in an official `script`

File: scripts/htmlDependencies.R [hunk 9]
```diff
@@ -19,7 +19,7 @@ withr::local_temp_libpaths()
 ignore <- capture.output({
   pak::pkg_install(c(
     "rstudio/bslib@main",
-    "rstudio/shiny@main",
```
Comment (cpsievert) [inline scripts/htmlDependencies.R:22]: TODO: get https://github.com/rstudio/shiny/pull/4040 merged and switch this back to `main`

**Comment by gfleetwood** [comment]
Hello. I've been tracking this thread for this feature, and tried it out - after upgrading the library - since I saw that it was merged. However, the line `ui.busy_indicators.options(spinner_type = 'ring')` gives an error:

module 'shiny.ui' has no attribute 'busy_indicators'

 When will this be added to the pypi package, and to the documentation?

**Comment by jcheng5** [comment]
@gfleetwood You can install from main with `pip install git+https://github.com/posit-dev/py-shiny`. We merge to main much more frequently than we release to PyPI, it might be a couple of weeks or so before the next one.

**Comment by gfleetwood** [comment]
@jcheng5 Got it. Thanks!

**Comment by gfleetwood** [comment]
@nstrayer Hello. I looked at the three examples showing how this feature works, but I'm still confused. It was working for a while in my app, but now it's not, and I don't know what is wrong. This is a minimal example, except that I will have multiple functions to display many dataframes, each having its own busy indicator.

```
from shiny import App, reactive, render, req, ui
from shiny.ui import TagList, div, h3, head_content, tags
import pandas as pd 
import requests

def server(input, output, session):

    @render.data_frame
    @reactive.event(input.search_button)
    def get_df():

        server_endpoint = 'https://'
        payload_in = {'search_term': input.search_term().strip()}

        #ui.busy_indicators.options(spinner_type = 'ring'),
        response = requests.post(server_endpoint, json = payload_in)
        payload_out = response.json()['data']

        result = render.DataGrid(pd.DataFrame(json.loads(payload_out)), width = '1000px') 

        return(result)

app_ui = ui.page_fluid(

    #ui.busy_indicators.options(spinner_type = 'ring'),

    ui.input_text('search_term', 'Enter Search Term'),
    ui.input_action_button('search_button', 'Search'),
    ui.busy_indicators.use(spinners = "spinners"),
    ui.output_data_frame('get_df'),
)

app = App(app_ui, server)
```

**Comment by cpsievert** [comment]
@gfleetwood there was actually a problem in the 0.10.1 that is now fixed in 0.10.2, please try upgrading to see if fixes your problems: `pip install -U shiny`

If you are still having problems after upgrading, please feel free to file a new issue (I promise I'll respond 😄 )

**Comment by gfleetwood** [comment]
@cpsievert No need, the upgrade did it. Thanks!

**Pull request merged**